creusot-contracts/src/local_invariant.rs

```diff
@@ -0,0 +1,224 @@
+//!         })
+//!     }
+//! }
+//! ```
+//!
+//! # Explicit namespaces
+//!
+//! For now, [`Namespaces`] must be explicitely passed to [`open`](LocalInvariant::open).
+//! We plan to relax this limitation at some point.
+
+use crate::{logic::Set, *};
+use ::std::cell::UnsafeCell;
+
+/// Declare a new namespace.
+///
+/// # Example
+///
+/// ```rust
+/// use creusot_contracts::{*, local_invariant::{declare_namespace, Namespace}, logic::Set};
+/// declare_namespace! { A }
+///
+/// #[requires(ns.contains(A()))]
+/// fn foo(ns: Ghost<&mut Set<Namespace>>) { /* ... */ }
+/// ```
+pub use base_macros::declare_namespace;
+
+/// The type of _namespaces_ associated with local invariants.
+///
+/// Can be declared with the [`declare_namespace`] macro, and then attached to a local
+/// invariant when creating it with [`LocalInvariant::new`].
+#[cfg_attr(creusot, rustc_diagnostic_item = "namespace_ty")]
+pub struct Namespace(());
+
+impl Namespace {
+    /// Used by [`declare_namespace`].
+    #[logic]
+    #[open(self)]
+    #[requires(false)]
+    #[doc(hidden)]
+    pub fn new() -> Self {
+        Self(())
+    }
+}
+
+/// A collection of namespaces.
+///
+/// This is given at the start of the program, and must be passed along to [LocalInvariant::open].
+pub struct Namespaces<'a>(::std::marker::PhantomData<&'a mut Vec<Namespace>>);
+
+impl Namespaces<'_> {
+    /// Get the underlying set of namespaces.
+    ///
+    /// Also accessible via the [`view`](View::view) (`@`) operator.
+    #[logic]
+    #[open(self)]
+    #[trusted]
+    pub fn namespaces(self) -> Set<Namespace> {
+        dead
+    }
+
+    /// Reborrow the namespace, allowing it to be reused later.
+    ///
+    /// # Example
+    /// ```
+    /// # use creusot_contracts::{*, local_invariant::Namespaces};
+    /// fn foo(namespaces: Ghost<Namespaces>) {}
+    /// fn bar(namespaces: Ghost<Namespaces>) {}
+    /// fn baz(mut namespaces: Ghost<Namespaces>) {
+    ///     foo(ghost!(namespaces.reborrow()));
+    ///     bar(namespaces);
+    /// }
+    /// ```
+    #[ensures((*self).namespaces() == result.namespaces())]
+    #[ensures((^self).namespaces() == result.namespaces())]
+    #[trusted]
+    #[pure]
+    pub fn reborrow(&mut self) -> Namespaces {
+        Namespaces(::std::marker::PhantomData)
+    }
+
+    #[logic]
+    #[open]
+    pub fn contains(self, namespace: Namespace) -> bool {
+        self.namespaces().contains(namespace)
+    }
+}
+
+impl View for Namespaces<'_> {
+    type ViewTy = Set<Namespace>;
+    #[logic]
+    #[open]
+    fn view(self) -> Set<Namespace> {
+        self.namespaces()
+    }
+}
+
+/// A ghost structure, that holds a piece of data (`T`) together with an
+/// [invariant](LocalInvariantSpec).
+pub struct LocalInvariant<T: LocalInvariantSpec> {
+    value: UnsafeCell<T>,
+    #[cfg_attr(not(creusot), allow(unused))]
+    public: Snapshot<T::Public>,
+    #[cfg_attr(not(creusot), allow(unused))]
+    namespace: Snapshot<Namespace>,
+}
```

Comment:
This really is a opaque (trusted) type, so it should be annotated as such, and therefore the fields `public` and `namespace` are not needed.

```suggestion
#[trusted]
pub struct LocalInvariant<T: LocalInvariantSpec> {
    value: UnsafeCell<T>,
}
```